### ISSUE TITLE
Lint suggestion: Replace `.chars().enumerate()` with `.char_indices()` when index is used in subsequent string slice operations

### ISSUE THREAD

**Issue by messense** [rust-lang/rust-clippy#10202]
### What it does

Suggest instances where `.chars().enumerate()` should be replaced with `.char_indices()` instead for str slices when the index returned by `.enumerate()` is used in subsequent string slice operations like `s[i..]` or `s.split_at(i)`.

Here is a real world example of this kind of issue: https://github.com/RustPython/RustPython/pull/4444, failure to correctly handle multi-byte character when doing string slice operations leads to panic at runtime.

### Category

correctness

### Advantage

- Correctly handle multi-byte character in string slice operations

### Example

```rust
let s = "{a:%ЫйЯЧ}";
for (i, c) in s.chars().enumerate() {
    let (_left, _right) = s.split_at(i);
}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2ce363c9482cb1a608913e50b6ed9bcb)
